// gp/lib/loader.js
var _ = require('underscore');
var request = require('request');
var Class = require('jc');

module.exports = Class({
	
	init: function (options) {
		var self = this;
		self.pages = {};
		self.queue = [];
		self.timeout = options.timeout || 3000;
		self.started = false;
	},
	
	get: function (url, callback) {
		var self = this;
		if (self.pages[url]) {
			callback(self.pages[url]);
			return;
		}
		self.queue.push({
			url: url,
			callback: callback
		});
		self.start();
	},
	
	start: function () {
		var self = this;
		if (self.started) { return; }
		self.started = true;
		next();
		
		function next() {
			if (self.queue.length === 0) {
				self.started = false;
				return;
			}
			var q = self.queue.shift();
			self.loadPage(q.url, function (content) {
				q.callback.call(null, content);
				wait();
			});
		}
		
		function wait() {
			setTimeout(function () {
				next();
			}, self.timeout);
		}
	},
	
	loadPage: function (url, callback) {
		var self = this;
		request(url, function (err, response, body) {
			self.pages[url] = body + '';
			callback(body + '');
		});
	},
	
});
